### PR TITLE
Various fixes to wwvegas codebase

### PR DESCRIPTION
**Pull request by jonwil**
Add DX8Wrapper::Clamp_Color.
Properly implement W3dUtilityClass
Add some things to w3d_file.h
Add TextureClass::Get_Level_Description
Add Greatest_Common_Divisor and Least_Common_Multiple
Add NodeCompressedMotionStruct constructor
Add SimpleFileFactoryClass and defaultFileFactory
Add DX8Wrapper::Convert_Color_Clamp

### CODE REVIEW

File: src/w3d/renderer/hcanim.h [hunk 1]
```diff
@@ -45,6 +45,7 @@ struct NodeCompressedMotionStruct
         } vd;
     };
     TimeCodedBitChannelClass *Vis;
+    NodeCompressedMotionStruct() : Vis(0), tc() {}
```
Comment (duncanspumpkin) [inline src/w3d/renderer/hcanim.h]: Vis(nullptr)

File: src/w3d/renderer/texture.cpp [hunk 2]
```diff
@@ -457,3 +457,13 @@ unsigned TextureClass::Get_Texture_Memory_Usage()
 
     return usage;
 }
+
+void TextureClass::Get_Level_Description(SurfaceClass::SurfaceDescription &surface_desc, unsigned int level)
+{
+    SurfaceClass *surface = Get_Surface_Level(level);
+
+    if (surface) {
```
Comment (duncanspumpkin) [inline src/w3d/renderer/texture.cpp]: != nullptr

File: src/w3d/renderer/hcanim.h (hunk 1 above)
Comment (jonwil) [reply src/w3d/renderer/hcanim.h]: Fixed.

File: src/w3d/renderer/w3d_util.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+void W3dUtilityClass::Convert_Vector(const IOVector3Struct &v, Vector3 *set)
+{
+    set->X = v.x;
+    set->Y = v.y;
+    set->Z = v.z;
+}
+
+void W3dUtilityClass::Convert_Color(const W3dRGBStruct &rgb, Vector3 *set)
+{
+    set->X = (float)rgb.r / 255;
```
Comment (duncanspumpkin) [inline src/w3d/renderer/w3d_util.cpp:30]: I saw this before and wondered why the `(float)`? You are dividing by a integer and storing in a float.

File: src/w3d/renderer/w3d_util.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+struct IOVector3Struct;
+class ShaderClass;
+class Vector4;
+class Vector3;
+struct W3dRGBAStruct;
+struct W3dRGBStruct;
+struct W3dShaderStruct;
+
+class W3dUtilityClass
```
Comment (duncanspumpkin) [inline src/w3d/renderer/w3d_util.h:26]: If all the members are static and public we are just using this as a namespace so may as well make it `namespace W3dUtilityClass`

File: src/w3d/renderer/w3d_util.cpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+{
+    set->X = (float)rgb.R / 255;
+    set->Y = (float)rgb.G / 255;
+    set->Z = (float)rgb.B / 255;
+    set->W = (float)rgb.A / 255;
+}
+
+void W3dUtilityClass::Convert_Shader(const W3dShaderStruct &shader, ShaderClass *set)
+{
+    set->Set_Depth_Compare((ShaderClass::DepthCompareType)W3d_Shader_Get_Depth_Compare(&shader));
```
Comment (duncanspumpkin) [inline src/w3d/renderer/w3d_util.cpp:45]: Is there no way these casts can be moved up into the Get functions.

File: src/w3d/lib/ffactory.cpp [hunk 6]
```diff
@@ -40,3 +45,67 @@ void RawFileFactoryClass::Return_File(FileClass *file)
 {
     delete file;
 }
+
+SimpleFileFactoryClass::SimpleFileFactoryClass() : m_isStripPath(false) {}
+
+bool Is_Full_Path(const char *path)
+{
+    if (!path) {
```
Comment (duncanspumpkin) [inline src/w3d/lib/ffactory.cpp]: `== nullptr`

File: src/w3d/lib/ffactory.cpp [hunk 7]
```diff
@@ -40,3 +45,67 @@ void RawFileFactoryClass::Return_File(FileClass *file)
...
+        return false;
+    }
+
+    return path[1] == ':' || (path[0] == '\\' && path[1] == '\\');
```
Comment (duncanspumpkin) [inline src/w3d/lib/ffactory.cpp:57]: Could add a note that this is `std::filesystem::path::has_root_name()`

File: src/w3d/lib/ffactory.cpp [hunk 8]
```diff
@@ -40,3 +45,67 @@ void RawFileFactoryClass::Return_File(FileClass *file)
...
+}
+
+FileClass *SimpleFileFactoryClass::Get_File(const char *filename)
+{
+    StringClass stripped_name(248, true);
+
+    if (m_isStripPath) {
+        const char *separator = strrchr(filename, '\\');
+
+        if (separator) {
```
Comment (duncanspumpkin) [inline src/w3d/lib/ffactory.cpp]: != nullptr

File: src/w3d/renderer/w3d_util.cpp (hunk 5 above)
Comment (jonwil) [reply src/w3d/renderer/w3d_util.cpp:45]: Nope.

File: src/w3d/renderer/texture.cpp (hunk 2 above)
Comment (jonwil) [reply src/w3d/renderer/texture.cpp]: Fixed.

File: src/w3d/renderer/w3d_util.h (hunk 4 above)
Comment (jonwil) [reply src/w3d/renderer/w3d_util.h:26]: Its a class in the original and I see no reason not to make it a class in Thyme at this point for consistency.

File: src/w3d/lib/ffactory.cpp (hunk 8 above)
Comment (jonwil) [reply src/w3d/lib/ffactory.cpp]: Fixed.

File: src/w3d/lib/ffactory.cpp (hunk 6 above)
Comment (jonwil) [reply src/w3d/lib/ffactory.cpp]: Fixed.

File: src/w3d/renderer/w3d_util.cpp (hunk 3 above)
Comment (jonwil) [reply src/w3d/renderer/w3d_util.cpp:30]: The float conversion ensures the / operator does a float divide and not an int divide

Comment (duncanspumpkin) [reply src/w3d/renderer/w3d_util.cpp:30]: then why not change the constant to `255.0f`

File: src/w3d/lib/gcd_lcm.cpp [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#include "gcd_lcm.h"
+
+unsigned int Greatest_Common_Divisor(unsigned int a,unsigned int b)
+{
+	unsigned int ret = a;
+	if (b)
+	{
```
Comment (duncanspumpkin) [inline src/w3d/lib/gcd_lcm.cpp]: you've tabs instead of spaces here causing the ci to fail.

File: src/w3d/renderer/dx8wrapper.h [hunk 10]
```diff
@@ -714,6 +716,21 @@ inline unsigned int DX8Wrapper::Convert_Color(const Vector4 &color)
...
+    color.Y = std::clamp(color.Y, 0.f, 1.f);
+    color.Z = std::clamp(color.Z, 0.f, 1.f);
+    color.W = std::clamp(color.W, 0.f, 1.f);
+}
+
+inline unsigned int DX8Wrapper::Convert_Color_Clamp(const Vector4 &color)
+{
+    Vector4 clamped_color = color;
+    DX8Wrapper::Clamp_Color(clamped_color);
+    return Convert_Color(reinterpret_cast<const Vector3 &>(clamped_color), clamped_color[3]);
```
Comment (duncanspumpkin) [inline src/w3d/renderer/dx8wrapper.h]: You can just pass it the vector4 there is already an overload for that and avoids the UB of reinterpret cast to convert a vec4 to a vec3.

**Review by duncanspumpkin — APPROVED** [review]
Looks good to me. Checked against wb for the simple file factory class.

**Pull request merged**